### PR TITLE
CONTRACTS: store contracts in dedicated symbols

### PR DESCRIPTION
**Pull request by tautschnig**
Contracts can now be placed on declarations or definitions, and result
in a fresh symbol being generated. That new symbol has its name prefixed
with "contracts::" to place them in a pseudo-namespace, and has
"is_property" set (which was an otherwise unused symbolt flag).

Co-authored-by: Remi Delmas <delmasrd@amazon.com>

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/ansi-c/c_typecheck_base.cpp
Comment (remi-delmas-3000) [inline src/ansi-c/c_typecheck_base.cpp:883]: Will the contract symbol also contain the function body if one was given by the user ? 

If possible I would like the contract symbol to have an empty body but to keep the contract parameters in the symbol table, because in a second step (goto-convert maybe ?) I would like to generate an executable body from the contract clauses.

File: src/ansi-c/c_typecheck_base.cpp
Comment (remi-delmas-3000) [inline src/ansi-c/c_typecheck_base.cpp]: Here I would rather replace the contract parameters with fresh symbols (maybe just by prefixing the function parameters with a `contract::`  prefix, and also rewrite the contract clauses with these new parameters using a `replacet`. 

Later we would be able to generate a body for the contract without symbol clashes between function and contract symbols and make the contract fully self-contained.

File: src/goto-instrument/contracts/contracts.cpp
Comment (remi-delmas-3000) [inline src/goto-instrument/contracts/contracts.cpp:756]: How can this equality work if the function symbol had its contract clauses removed but the contract symbol still has them ? 

Would the check still work if in addition we removed the function body from the contract and renamed contract parameters ? 

Maybe we could just check that the number of arguments and their types are equal.

File: src/linking/remove_internal_symbols.cpp
Comment (remi-delmas-3000) [inline src/linking/remove_internal_symbols.cpp:217]: If a function `foo` is file_local, is also `contract::foo` file_local ? 

What if some of the contract's clauses use file_local symbols in preconditions or post conditions or assigns clause ?

**Review by remi-delmas-3000 — COMMENTED** [review]
It would be nice to make the contract completely self-contained by introducing fresh/prefixed names for the parameters, rewriting the contract clauses with the fresh names, and removing the body if any.

File: src/ansi-c/c_typecheck_base.cpp
Comment (tautschnig) [reply src/ansi-c/c_typecheck_base.cpp:883]: No, the contract symbol _at present_ really just has the type information including the contract annotation. We could obviously evolve that.

I'm not sure whether symbols for the contract parameters are required to remain in the symbol table: even the type signature already has the parameters with their names and their types, so we can easily create symbols from them if need be.

File: src/ansi-c/c_typecheck_base.cpp
Comment (tautschnig) [reply src/ansi-c/c_typecheck_base.cpp]: I believe we could still choose to create those fresh symbols if and when needed: all the required information still is in the contract symbol.

File: src/goto-instrument/contracts/contracts.cpp
Comment (tautschnig) [reply src/goto-instrument/contracts/contracts.cpp:756]: This perhaps warrants an extra comment: the comparison works, because all the extra information is stored in comment-elements of the ireps: both the parameter identifiers as well as all the contract clauses have `ID_C_...` keys in the `irept`. Therefore, equality (but not full equality!) works out. While the body isn't there as indicated in the earlier comment, it wouldn't affect this comparison either way for it's just a comparison of the types, not the symbols' values.

File: src/linking/remove_internal_symbols.cpp
Comment (tautschnig) [reply src/linking/remove_internal_symbols.cpp:217]: No, the contract does not currently have any flags (other than `is_property` to mark it as a contract) set, and will therefore not be considered "file_local."

Your call-out about file_local symbols is a good one as the linker could choose to rename them in case of a conflict, at which point the contract would no longer refer to the correct symbol. I'll try to come up with some code that tracks this renaming and adjusts the contract accordingly.

File: src/ansi-c/c_typecheck_base.cpp
Comment (kroening) [reply src/ansi-c/c_typecheck_base.cpp:883]: Use a binding instead!

Comment (tautschnig) [reply src/ansi-c/c_typecheck_base.cpp:883]: > Use a binding instead!

For all that I can tell this requires some preparing work. https://github.com/diffblue/cbmc/pull/6811 will be a starting point.

File: src/linking/remove_internal_symbols.cpp
Comment (remi-delmas-3000) [reply src/linking/remove_internal_symbols.cpp:217]: Hi again,
I experimented with the following:


```c
// foo.h
#include <stdlib.h>
void foo(char *arr, size_t size);
```

```c
// foo.c
#include "foo.h"
void foo(char *arr, size_t size) {
    if(arr && size > 0)
        arr[0] = 0;
}
```

```c
// foo_contract.c
#include <stdlib.h>
void foo(char *arr, size_t size)
__CPROVER_assigns(arr && size > 0 : arr[0]);
```

```c
// main.c
#include "foo.h"
void main() {
    size_t size = 10;
    char *arr = malloc(size);
    foo(arr, size);
}
```

and then 

```bash
▶goto-cc foo_contract.c foo.c main.c  -o project.o && goto-instrument project.o project-mod.o --enforce-contract foo && cbmc project-mod.o

** Results:
<builtin-library-malloc> function malloc
[malloc.assertion.1] line 26 max allocation size exceeded: SUCCESS
[malloc.assertion.2] line 31 max allocation may fail: SUCCESS

foo.c function foo
[foo.assigns.2] line 4 Check that arr[(signed long int)0] is assignable: SUCCESS
[foo.assigns.1] file foo_contract.c line 4 Check that arr[(signed long int)0] is valid when arr != ((char *)NULL) && size > (unsigned long int)0: SUCCESS

** 0 of 4 failed (1 iterations)
VERIFICATION SUCCESSFUL
```

But if I include `"foo.h"` in `"foo_contract.h"` like so

```c
// foo_contract.c
#include <stdlib.h>
#include <foo.h>
void foo(char *arr, size_t size)
__CPROVER_assigns(arr && size > 0 : arr[0]);
```


Then it seems like the contract clauses are lost.

```bash
** Results:
<builtin-library-malloc> function malloc
[malloc.assertion.1] line 26 max allocation size exceeded: SUCCESS
[malloc.assertion.2] line 31 max allocation may fail: SUCCESS

foo.c function foo
[foo.assigns.1] line 4 Check that arr[(signed long int)0] is assignable: FAILURE

** 1 of 3 failed (2 iterations)
VERIFICATION FAILED
```

Comment (tautschnig) [reply src/linking/remove_internal_symbols.cpp:217]: Good catch, we need to maintain the contract in `typecheck_redefinition_non_type` even when the new symbol does not provide the function definition. I'll add this case and will try to include a test.

Comment (tautschnig) [reply src/linking/remove_internal_symbols.cpp:217]: Thanks @remi-delmas-3000, I believe I have now fixed this problem (I have also extended the test suite to capture this scenario).

File: src/ansi-c/c_typecheck_base.cpp
Comment (tautschnig) [reply src/ansi-c/c_typecheck_base.cpp:883]: I have now added a commit that wraps contracts in a `lambda_exprt` and uses `.application()` instead of resorting to `replace_symbolt`.

**Comment by tautschnig** [comment]
@kroening @remi-delmas-3000 I have put the additional changes to make use of `lambda_exprt` in a commit of its own, but it should now be working.

File: src/ansi-c/c_typecheck_base.cpp
Comment (feliperodri) [inline src/ansi-c/c_typecheck_base.cpp:382]: nitpick: can we cover this error with a regression test?

File: src/goto-instrument/contracts/contracts.cpp
Comment (feliperodri) [inline src/goto-instrument/contracts/contracts.cpp:760]: nitpick: could we cover this with a regression test?

**Review by feliperodri — APPROVED** [review]
LGTM, thanks @tautschnig!

File: src/goto-instrument/contracts/contracts.cpp
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/contracts.cpp]: [nit] could use `auto` here
```suggestion
  const auto function_str = id2string(function);
  const auto &function_symbol = ns.lookup(function);
```

File: src/goto-instrument/contracts/contracts.cpp
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/contracts.cpp:1454]: Thanks for cleaning this up. We don't seem to be using this `skip` target any more.
While you're at it, could you please also update the pseudocode in the comments above to reflect the current instrumentation?

**Review by SaswatPadhi — APPROVED** [review]
LGTM overall; couple of minor comments below.

**Comment by feliperodri** [comment]
@chris-ryder, @kroening, @martin-cs, and/or @peterschrammel, could we get one more review on this PR? It's missing code owner approval.

**Review by peterschrammel — APPROVED** [review]

File: src/ansi-c/c_typecheck_base.cpp
Comment (tautschnig) [reply src/ansi-c/c_typecheck_base.cpp:382]: Done!

File: src/goto-instrument/contracts/contracts.cpp
Comment (tautschnig) [reply src/goto-instrument/contracts/contracts.cpp:760]: Done!

**Review by kroening — APPROVED** [review]

**Pull request merged**
